File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/gamePlay/SwitchPlayerAction.java
package nl.hva.backend.domain.models.game.action.impl.gamePlay;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.GameAction;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;
import java.time.LocalDateTime;

/**
 * TODO: Add JavaDoc
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class SwitchPlayerAction extends GameAction {

    @ManyToOne
    private Player newPlayer;

    private LocalDateTime turnEndsAt;

    public SwitchPlayerAction(
            Player actor,
            Player newPlayer,
            LocalDateTime turnEndsAt
    ) {
        super(actor);
        this.newPlayer = newPlayer;
        this.turnEndsAt = turnEndsAt;
    }

}

File: front-end/src/app/domain/models/game/action/impl/gamePlay/SwitchPlayerAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";

/**
 * TODO: Add JsDoc
 *
 * @author Hamza el Haouti
 */
export class SwitchPlayerAction extends GameAction {
  private _newPlayer: Player;
  private _turnEndsAt: Date;

  public constructor(
    id: number,
    actor: Player,
    performedAt: Date,
    newPlayer: Player,
    turnEndsAt: Date
  ) {
    super(SwitchPlayerAction.name, id, actor, performedAt);
    this._newPlayer = newPlayer;
    this._turnEndsAt = turnEndsAt;
  }

  public get newPlayer(): Player {
    return this._newPlayer;
  }

  public get turnEndsAt(): Date {
    return this._turnEndsAt;
  }

  public static trueCopy(action: SwitchPlayerAction): GameAction | null {
    return new SwitchPlayerAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      Player.trueCopy(action.newPlayer),
      new Date(action.turnEndsAt)
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "newPlayer": this.newPlayer.toJson(),
        "turnEndsAt": this.turnEndsAt.toJSON(),
      });
  }

  public override toString(): string {
    return `Turn of ${super.actor.user.username} ended. It's now the turn of ${this.newPlayer.user.username}`;
  }

}
